### PR TITLE
fix(repo-server): avoid fetching commit sha for multisource applications (#15037)

### LINKED ISSUE #15037: Annotated Tag's Commit SHA Not Resolved Properly
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Bug description**

When an application revision is defined by an annotated tag in a multisource application, and one or more of the referenced sources are fetched from the same repository as the main source, the repository validation fails. See the following YAML exampl:  
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  annotations:
    app.of.apps/managed-by: argo-cd
  labels:
    argocd.argoproj.io/instance: cluster-startup
  name: eck-orchestration-app
  namespace: argocd
spec:
  destination:
    namespace: eck-operator
    server: 'https://kubernetes.default.svc'
  project: default
  sources:
    - helm:
        releaseName: eck-orchestration-app
        valueFiles:
          - $global/values/global-values.yaml
          - $overlay/internal-feeds-dev/argocd/kibana-overlay-values.yaml
          - $overlay/internal-feeds-dev/argocd/elasticsearch-overlay-values.yaml
      path: elasticsearch
      repoURL: 'https://github.com/REDACTED/argocd-apps.git'
      targetRevision: v1.6.0
    - ref: overlay
      repoURL: 'https://github.com/REDACTED/conf.git'
      targetRevision: internal-feeds-dev-cluster
    - ref: global
      repoURL: 'https://github.com/REDACTED/argocd-apps.git'
      targetRevision: v1.6.0
```
The repository validation fails with an error defined in [repository.go file](https://github.com/argoproj/argo-cd/blob/f7bfb3f15b7d3e7524c705e6cf3d18659d14186e/reposerver/repository/repository.go#L730-L733).

The main reason for this failure is: [this line](https://github.com/argoproj/argo-cd/blob/f7bfb3f15b7d3e7524c705e6cf3d18659d14186e/reposerver/repository/repository.go#L402).

**Explanation**
The function on line 402 fetches the HEAD revision of the git repo, and this revision is different from the SHA of the [annotated tag](https://git-scm.com/book/en/v2/Git-Basics-Tagging), However, the validation on line 730 uses the referencedCommit revision, which equals the SHA of the annotated tag. Since the committedSHA and referencedCommitSHA are different, the validation fails.

**Suggestion to Fix**
Avoid using HEAD fetching for multisource applications in line 402.
Suggested code below:
```go
var commitSHA string
if !hasMultipleSources {
	commit, err := gitClient.CommitSHA()
	if err != nil {
		return err
	}
	commitSHA = commit
} else {
	commitSHA = revision
}
```

I'll be happy to discuss this fix, and if the community agrees, I will open a PR for this fix.

### PR DESCRIPTION
**Pull request by ozlevka-work**
Fixes #15037, #13720
Changes to avoid fetching commit SHA for multisource applications. Broke annotated tags targetRevisions

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

### CODE REVIEW

File: .github/workflows/ci-build.yaml [hunk 1]
```diff
@@ -76,7 +76,7 @@ jobs:
         with:
           go-version: ${{ env.GOLANG_VERSION }}
       - name: Run golangci-lint
-        uses: golangci/golangci-lint-action@639cd343e1d3b897ff35927a75193d57cfcba299 # v3.6.0
+        uses: golangci/golangci-lint-action@3a919529898de77ec3da873e3063ca4b10e7f5cc # v3.7.0
```
Comment (jannfis) [inline .github/workflows/ci-build.yaml:79]: Please remove unrelated changes from this PR.

File: .github/workflows/ci-build.yaml [hunk 2]
```diff
@@ -265,7 +265,7 @@ jobs:
       - name: Checkout code
         uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
       - name: Setup NodeJS
-        uses: actions/setup-node@e33196f7422957bea03ed53f6fbb155025ffc7b8 # v3.7.0
+        uses: actions/setup-node@bea5baf987ba7aa777a8a0b4ace377a21c45c381 # v3.8.0
```
Comment (jannfis) [inline .github/workflows/ci-build.yaml:268]: Please remove unrelated changes from this PR.

File: README.md [hunk 3]
```diff
@@ -56,7 +56,7 @@ Participation in the Argo CD project is governed by the [CNCF Code of Conduct](h
 ### Blogs and Presentations
 
 1. [Awesome-Argo: A Curated List of Awesome Projects and Resources Related to Argo](https://github.com/terrytangyuan/awesome-argo)
-1. [Unveil the Secret Ingredients of Continuous Delivery at Enterprise Scale with Argo CD](https://blog.akuity.io/unveil-the-secret-ingredients-of-continuous-delivery-at-enterprise-scale-with-argo-cd-7c5b4057ee49)
+1. [Unveil the Secret Ingredients of Continuous Delivery at Enterprise Scale with Argo CD](https://akuity.io/blog/unveil-the-secret-ingredients-of-continuous-delivery-at-enterprise-scale-with-argocd-kubecon-china-2021/)
```
Comment (jannfis) [inline README.md:59]: Please remove unrelated changes from this PR.

**Comment by jannfis** [comment]
Can you please rebase your PR? It contains a ton of unrelated changes. Thanks!

**Comment by jannfis** [comment]
Cool, thanks a bunch for the rebase. Can we back this change by a unit test, perhaps?

**Comment by ozlevka-work** [comment]
Hi @jannfis,
Thank you for your comment.
Rebase is done. Please review.

File: reposerver/repository/repository.go [hunk 4]
```diff
@@ -399,9 +399,15 @@ func (s *Service) runRepoOperation(
...
-		commitSHA, err := gitClient.CommitSHA()
-		if err != nil {
-			return err
+		var commitSHA string
+		if hasMultipleSources {
+			commitSHA = revision
+		} else {
+			commit, err := gitClient.CommitSHA()
+			if err != nil {
+				return err
```
Comment (crenshaw-dev) [inline reposerver/repository/repository.go:408]:
```suggestion
				return fmt.Errorf("failed to get commit SHA: %w", err)
```

Comment (ozlevka-work) [reply reposerver/repository/repository.go:408]: Thank you...

**Comment by tkaesserfm** [comment]
@crenshaw-dev can we move forward with this? Would really love to see this in upstream. Currently not having this prevents us to move forward with some features to production because we don't want to get an argocd alert from time to time and this would help :)

**Review by crenshaw-dev — APPROVED** [review]
Many thanks, @ozlevka-work!

**Comment by tkaesserfm** [comment]
@crenshaw-dev @ozlevka-work Thank you so much guys!
